src/compiler/crystal/semantic/restrictions.cr

```diff
@@ -32,40 +32,44 @@ require "../types"
 # `Int32 | String` against `Int32`, the result is `Int32`.
 
 module Crystal
+  record RestrictionsContext, owner : Type,
+    self_def : DefWithMetadata? = nil,
+    other_def : DefWithMetadata? = nil
+
   class ASTNode
-    def restriction_of?(other : Underscore, owner)
+    def restriction_of?(other : Underscore, ctx)
       true
     end
 
-    def restriction_of?(other : ASTNode, owner)
+    def restriction_of?(other : ASTNode, ctx)
       self == other
     end
 
-    def restriction_of?(other : Type, owner)
+    def restriction_of?(other : Type, ctx)
       false
     end
 
-    def restriction_of?(other, owner)
+    def restriction_of?(other, ctx)
       raise "BUG: called #{self}.restriction_of?(#{other})"
     end
   end
 
   class Self
-    def restriction_of?(type : Type, owner)
-      owner.restriction_of?(type, owner)
+    def restriction_of?(type : Type, ctx)
+      ctx.owner.restriction_of?(type, ctx)
     end
 
-    def restriction_of?(type : Self, owner)
+    def restriction_of?(type : Self, ctx)
       true
     end
 
-    def restriction_of?(type : ASTNode, owner)
+    def restriction_of?(type : ASTNode, ctx)
       false
     end
   end
 
   struct DefWithMetadata
-    def restriction_of?(other : DefWithMetadata, owner)
+    def restriction_of?(other : DefWithMetadata, ctx)
```

Comment:
I think this was fine before. We can create the context inside this method because we have everything: `self`, `other` and `ctx`. That means not creating the context in `ModuleType#add_def`.

Reply:
I thought about it but thought it would be confusing at call site or while reading the code, but I've found a way using a named arg for `DefWithMetadata#restriction_of?` to make it clear everywhere :smiley: